ENIGMAsystem/SHELL/Universal_System/Serialization/type_traits.h

```diff
@@ -0,0 +1,277 @@
+    I, std::enable_if_t<I<std::tuple_size_v<typename TupleTypeExtractor<T>::TupleTypes>>> {
+  using ValueType = std::tuple_element_t<I, typename TupleTypeExtractor<T>::TupleTypes>;
+  using RemainingTypes = TupleTypeExtractorHelper<T, I + 1>;
+  using TupleTypes = std::tuple<ValueType, typename RemainingTypes::TupleTypes>;
+  using ResultType = std::tuple<ValueType, typename RemainingTypes::ResultType>;
+};
+
+template <typename... Args>
+struct TupleSize;
+
+template <typename... Args>
+struct TupleSize<std::tuple<Args...>> {
+  static const std::size_t value;
+};
+
+template <typename... Args>
+const std::size_t TupleSize<std::tuple<Args...>>::value = sizeof...(Args);
+
+template <typename... Args>
+struct TupleSize<const std::tuple<Args...>&> {
+  static const std::size_t value;
+};
+
+template <typename... Args>
+const std::size_t TupleSize<const std::tuple<Args...>&>::value = sizeof...(Args);
+
+template <std::size_t Index, typename Tuple, typename Function, typename ExtraParam>
+struct TupleLooper {
+  static void loop(const Tuple& tuple, Function&& func, ExtraParam& extraParam) {
+    TupleLooper<Index - 1, Tuple, Function, ExtraParam>::loop(tuple, std::forward<Function>(func), extraParam);
+    func(std::get<Index - 1>(tuple), extraParam);
+  }
+};
+
+template <typename Tuple, typename Function, typename ExtraParam>
+struct TupleLooper<0, Tuple, Function, ExtraParam> {
+  static void loop(const Tuple& tuple, Function&& func, const ExtraParam& extraParam) {}
+};
+
+template <typename... Args, typename Function, typename ExtraParam>
+void LoopTuple(const std::tuple<Args...>& tuple, Function&& func, ExtraParam& extraParam) {
+  TupleLooper<sizeof...(Args), std::tuple<Args...>, Function, ExtraParam>::loop(tuple, std::forward<Function>(func),
+                                                                                extraParam);
+}
+
+template <typename T, std::size_t N>
+struct has_nested_form : std::false_type {
+  using inner_type = void;
+};
+
+template <typename T>
+struct has_nested_form<T, 0> : std::true_type {
+  using inner_type = T;
+};
+
+template <template <typename> typename T, typename U, std::size_t N>
+struct has_nested_form<T<U>, N> : has_nested_form<U, N - 1> {
+  using inner_type = U;
+};
+
+template <typename T, std::size_t N>
+constexpr static inline bool has_nested_form_v = has_nested_form<T, N>::value;
+
+template <typename T>
+constexpr static inline bool always_false = false;
+
+template <typename T, typename ExpectedType, typename ReturnType = void>
+using is_t = std::enable_if_t<std::is_same_v<ExpectedType, std::decay_t<T>>, ReturnType>;
+
+template <typename... Conds>
+struct _logical_or_ : std::false_type {};
+
+template <typename Cond, typename... Rest>
+struct _logical_or_<Cond, Rest...> : std::conditional_t<Cond::value, std::true_type, _logical_or_<Rest...>> {};
+
+template <typename... Conds>
+constexpr static inline bool _logical_or_v = _logical_or_<Conds...>::value;
```

Comment:
I do believe this can be replaced by [`std::disjunction`](https://en.cppreference.com/w/cpp/types/disjunction).